### PR TITLE
NXP-29865: fix getUsersInGroupAndSubGroups method of the UserManager

### PR DESCRIPTION
**Pull request by charlesboidot**

### CODE REVIEW

**Review by troger — CHANGES_REQUESTED** [review]
A test needs to be added, probably on the public `UserManager#getUsersInGroupAndSubGroups` methods to validate the use case explain in the NXP ticket.

Moreover, you can rework the commit message as the change is not directly related to JSF or notifications but to the `UserManager#getUsersInGroupAndSubGroups` behavior IMO.

**Comment by charlesboidot** [comment]
A test already existed for this method so I just changed it so it would test the modifications made.

File: modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java [hunk 1]
```diff
@@ -596,34 +596,46 @@ public void testGetUsersInGroup() throws Exception {
...
+        g1.setProperty("group", "subGroups", g1Groups);
         userManager.createGroup(g1);
         g2.setProperty("group", "members", g2Users);
         g2.setProperty("group", "subGroups", g2Groups);
         userManager.createGroup(g2);
+        g3.setProperty("group", "members", g3Users);
+        userManager.createGroup(g3);
 
-        List<String> expectedUsersInGroup1 = Collections.singletonList("test_u1");
+        List<String> expectedUsersInGroup1 = Arrays.asList("test_u1", "test_u3");
```
Comment (troger) [inline modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java:631]:
```suggestion
        List<String> expectedUsersInGroup1 = List.of("test_u1", "test_u3");
```
(but you probably need to use `Arrays.asList` on the 10.10 backport)

**Review by troger — APPROVED** [review]

File: modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java [hunk 2]
```diff
@@ -1440,18 +1452,19 @@ protected void doTestGetQueryForPattern(boolean compat) throws Exception {
 
         // basic query, directory configured for "subinitial" substring match type
         query = getQueryForPattern("foo");
-        assertEquals("((firstName ILIKE 'foo%') OR (lastName ILIKE 'foo%') OR (username ILIKE 'foo%'))",
-                query);
+        assertEquals("((firstName ILIKE 'foo%') OR (lastName ILIKE 'foo%') OR (username ILIKE 'foo%'))", query);
```
Comment (efge) [inline modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java:1455]: Reminder for next time: please avoid reformatting code in the same commit as other regular changes, it slows down review.

**Review by efge — APPROVED** [review]

File: modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java (hunk 2 above)
Comment (charlesboidot) [reply modules/platform/nuxeo-platform-usermanager/src/test/java/org/nuxeo/ecm/platform/usermanager/TestUserManager.java:1455]: I'll keep it in mind for next time, thanks for the reminder

File: modules/platform/nuxeo-platform-usermanager/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java [hunk 3]
```diff
@@ -685,7 +685,7 @@ protected NuxeoGroup makeGroup(DocumentModel groupEntry) {
     }
 
     protected void appendSubgroups(String groupId, Set<String> groups, DocumentModel context) {
-        List<String> groupsToAppend = getGroupsInGroup(groupId, context);
+        List<String> groupsToAppend = getDescendantGroups(groupId);
         groups.addAll(groupsToAppend);
         for (String subgroupId : groupsToAppend) {
             groups.add(subgroupId);
```
Comment (efge) [inline modules/platform/nuxeo-platform-usermanager/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java:691]: Sorry I missed this in the first round of review but the rest of the method seems completely redundant now. Could you check?

**Review by efge — CHANGES_REQUESTED** [review]

File: modules/platform/nuxeo-platform-usermanager/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java (hunk 3 above)
Comment (charlesboidot) [reply modules/platform/nuxeo-platform-usermanager/src/main/java/org/nuxeo/ecm/platform/usermanager/UserManagerImpl.java:691]: Indeed it doesn't do anything, I'm going to update the PRs

**Review by efge — DISMISSED** [review]

**Review by troger — DISMISSED** [review]

**Comment by charlesboidot** [comment]
Just rebased on master.

**Review by troger — APPROVED** [review]

**Review by efge — APPROVED** [review]

**Pull request merged**